src/lib/backoff.ts

```diff
@@ -0,0 +1,88 @@
+import { promisify } from 'util';
+
+const sleep = promisify(setTimeout);
+
+export type BackoffCallback = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: BackoffCallback) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
```

Comment:
Although is not necessary with typescript, it would not hurt to still assert that the passed argument is a function


```
import { strict as assert } from 'assert';

export function withBackoff.... {
   assert(typeof func === 'function', 'expected a function as parameter');
}
```